tests/rustdoc/hidden-private.rs

```diff
@@ -0,0 +1,42 @@
+// This is a regression test for <https://github.com/rust-lang/rust/issues/106373>.
+// It ensures that the items in the `doc(hidden)` const block don't show up in the
+// generated docs.
+
+// compile-flags: --document-private-items
+
+#![crate_name = "foo"]
+
+// @has 'foo/index.html'
+// @count - '//*[@class="item-table"]//a[@class="struct"]' 2
+// @count - '//*[@class="item-table"]//a[@class="trait"]' 1
+#[doc(hidden)]
+const _: () = {
+    macro_rules! stry {
+        () => {};
+    }
+
+    struct ShouldBeHidden;
+
+    // @has 'foo/struct.Foo.html'
+    // @!has - '//*[@class="code-header"]' 'impl Bar for Foo'
+    #[doc(hidden)]
+    impl Bar for Foo {
+        fn bar(&self) {
+            struct SHouldAlsoBeHidden;
+        }
+    }
+
+    // @has 'foo/struct.Private.html'
+    // @!has - '//*[@id="impl-Bar-for-Private"]/*[@class="code-header"]' 'impl Bar for Private'
```

Comment:
Why is this hidden? It means that If have a private type, and use `#[derive(Serialize)]`, the impl won't appear with `--document-private-items`, which seems undesirable.

Reply:
This is exactly the discussion we had [here](https://github.com/rust-lang/rust/pull/107000#discussion_r1072750801). If something is inside a hidden item's descendants, then it doesn't appear (not private, hidden).